DataFormats/Provenance/interface/CompactEventAuxiliaryVector.h

```diff
@@ -0,0 +1,188 @@
+        return details::hash_value(
+            extra.processHistoryID_.compactForm(), extra.isRealData_, extra.experimentType_, extra.storeNumber_);
+      }
+    };
+
+    using GUIDmemo = std::unordered_set<std::string>;
+    using extraMemo = std::unordered_set<CompactEventAuxiliaryExtra, ExtraHash>;
+
+    class CompactEventAuxiliary {
+    public:
+      CompactEventAuxiliary(EventID const& theId,
+                            std::string const& processGUID,
+                            Timestamp const& theTime,
+                            int bunchXing,
+                            int orbitNum,
+                            CompactEventAuxiliaryExtra const& extra,
+                            GUIDmemo& guidmemo,
+                            extraMemo& extramemo)
+          : id_(theId),
+            processGUID_(memoize(processGUID, guidmemo)),
+            time_(theTime),
+            bunchCrossing_(bunchXing),
+            orbitNumber_(orbitNum),
+            extra_(memoize(extra, extramemo)) {}
+      CompactEventAuxiliary(const EventAuxiliary& ea, GUIDmemo& guidmemo, extraMemo& extramemo)
+          : id_(ea.id()),
+            processGUID_(memoize(ea.processGUID(), guidmemo)),
+            time_(ea.time()),
+            bunchCrossing_(ea.bunchCrossing()),
+            orbitNumber_(ea.orbitNumber()),
+            extra_(memoize(CompactEventAuxiliaryExtra(ea), extramemo)) {}
+
+      void write(std::ostream& os) const;
+
+      ProcessHistoryID const& processHistoryID() const { return extra_.processHistoryID_; }
+      EventID const& id() const { return id_; }
+      std::string const& processGUID() const { return processGUID_; }
+      Timestamp const& time() const { return time_; }
+      LuminosityBlockNumber_t luminosityBlock() const { return id_.luminosityBlock(); }
+      EventNumber_t event() const { return id_.event(); }
+      RunNumber_t run() const { return id_.run(); }
+      bool isRealData() const { return extra_.isRealData_; }
+      ExperimentType experimentType() const { return extra_.experimentType_; }
+      int bunchCrossing() const { return bunchCrossing_; }
+      int orbitNumber() const { return orbitNumber_; }
+      int storeNumber() const { return extra_.storeNumber_; }
+
+      EventAuxiliary eventAuxiliary() const {
+        auto ea{EventAuxiliary(id_,
+                               processGUID_,
+                               time_,
+                               extra_.isRealData_,
+                               extra_.experimentType_,
+                               bunchCrossing_,
+                               extra_.storeNumber_,
+                               orbitNumber_)};
+        ea.setProcessHistoryID(extra_.processHistoryID_);
+        return ea;
+      }
+
+    private:
+      template <typename T, typename C>
+      const T& memoize(const T& item, C& memopad) const {
+        auto it = memopad.insert(item);
+        return *it.first;
+      }
+
+      // Event ID
```

Comment:
Given we are trying to save memory, how about re-arranging the order of the member data to avoid padding? The references have to be 64bit aligned so how about moving them to be first. Timestamp has two uints and EventID has two uints and a u long long (in that order). So looks like EventID should follow the references, then Timestamp and finally the ints.

Reply:
Never mind, I played around on godbolt and it didn't make a difference.

https://godbolt.org/z/r8ac1s

Reply:
I did check that there were an even number of 32-bit ints, so the 64-bit pointers are aligned without padding